File: src/agent/tools/builtin/ask_user.py
"""AskUser tool: prompt the user for clarification mid-conversation."""

from __future__ import annotations

from typing import Any

from agent.cli import CLI


class AskUserTool:
    """Ask the user a question when clarification is needed."""

    def __init__(self, cli: CLI) -> None:
        self._cli = cli

    @property
    def name(self) -> str:
        return "ask_user"

    @property
    def description(self) -> str:
        return (
            "Ask the user a question when you need clarification before proceeding. "
            "Use this when the request is ambiguous, you need to choose between multiple "
            "approaches, or you need information you can't determine from the codebase."
        )

    @property
    def input_schema(self) -> dict[str, Any]:
        return {
            "type": "object",
            "properties": {
                "question": {
                    "type": "string",
                    "description": "The question to ask the user.",
                },
            },
            "required": ["question"],
        }

    async def execute(self, *, question: str, **_: Any) -> str:
        self._cli.console.print(f"\n[bold cyan]? {question}[/bold cyan]")
        answer = await self._cli.get_input()
        if answer is None:
            return "(user did not provide an answer)"
        return answer

File: src/tools/builtin/askUser.ts
/** AskUser tool: prompt the user for clarification mid-conversation. */

import type { CLI } from '../../cli/types.js';
import type { Tool } from '../base.js';

export class AskUserTool implements Tool {
  constructor(private cli: CLI) {}

  get name(): string {
    return 'ask_user';
  }

  get description(): string {
    return (
      'Ask the user a question when you need clarification before proceeding. ' +
      'Use this when the request is ambiguous, you need to choose between multiple ' +
      "approaches, or you need information you can't determine from the codebase."
    );
  }

  get inputSchema(): Record<string, any> {
    return {
      type: 'object',
      properties: {
        question: {
          type: 'string',
          description: 'The question to ask the user.',
        },
      },
      required: ['question'],
    };
  }

  async execute(params: Record<string, any>): Promise<string> {
    const question: string = params.question;
    this.cli.printQuestion(question);
    const answer = await this.cli.getInput();
    if (answer === null) {
      return '(user did not provide an answer)';
    }
    return answer;
  }
}
